### PR TITLE
MGMT-12755: Disallow periods in cluster name

### PR DESCRIPTION
**Pull request by CrystalChun**
[MGMT-12755](https://issues.redhat.com/browse/MGMT-12755)
Cluster names containing periods "." are not allowed
by the openshift installer when using a platform that's
not the None platform.

This prevents a cluster name from passing the validation
if it contains any periods and if it's not the None platform.

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [x] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [x] assisted-test-infra environment: install a cluster with a period in the name (export CLUSTER_NAME=test.cluster) and see that it fails (check the service logs for error message:
```
Failed to registered cluster
```

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by CrystalChun** [comment]
Subdomains were originally allowed https://issues.redhat.com/browse/MGMT-11614
Going to investigate and verify before reopening this PR

File: internal/cluster/validations/validations.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ type Config struct {
 }
 
 const (
-	clusterNameRegex    = "^([a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*)*$"
+	clusterNameRegex    = "^([a-z0-9]([-a-z0-9]*[a-z0-9])?([a-z0-9]([-a-z0-9]*[a-z0-9])?)*)*$"
```
[the remark this answers is not in the document]
Comment (CrystalChun) [reply internal/cluster/validations/validations.go:38]: Good point, thanks Daniel!

**Comment by CrystalChun** [comment]
The openshift installer now doesn't allow periods when the cluster is:

- vsphere
- baremetal
- openstack
- nutanix

(more details at bottom)

---
Solution ideas:
1. Propagate the same validation to the service (don't allow periods when cluster is any of the 4 platforms above)
2. Don't allow any periods regardless of platform (reverts the subdomain requirement [MGMT-11614](https://issues.redhat.com/browse/MGMT-11614))
3. Keep the validation as is and let the openshift installer show the failure

---

Openshift installer change was added in this PR https://github.com/openshift/installer/pull/5931

https://github.com/rna-afk/installer/blob/780b29adbe5ba47a1b41ffeade875583282995db/pkg/types/validation/installconfig.go#L93

File: internal/cluster/validations/validations.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (CrystalChun) [reply internal/cluster/validations/validations.go:38]: Depending on the chosen solution, I'll revert it back to this if we're doing the old validation!

Do you have any thoughts about what we should do? https://github.com/openshift/assisted-service/pull/4778#issuecomment-1351589984

**Comment by danielerez** [comment]
> The openshift installer now doesn't allow periods when the cluster is:
> 
> * vsphere
> * baremetal
> * openstack
> * nutanix
> 
> (more details at bottom)
> 
> Solution ideas:
> 
> 1. Propagate the same validation to the service (don't allow periods when cluster is any of the 4 platforms above)
> 2. Don't allow any periods regardless of platform (reverts the subdomain requirement [MGMT-11614](https://issues.redhat.com/browse/MGMT-11614))
> 3. Keep the validation as is and let the openshift installer show the failure
> 
> Openshift installer change was added in this PR [openshift/installer#5931](https://github.com/openshift/installer/pull/5931)
> 
> https://github.com/rna-afk/installer/blob/780b29adbe5ba47a1b41ffeade875583282995db/pkg/types/validation/installconfig.go#L93

Oh nice catch, so it's a limitation of the platforms then. So maybe we could just enable the subdomain requirement only for none platform for now? I.e. It could be safer to have a stricter validation to avoid this issue when adding more platforms later on. Do we know whether the subdomain requirement indeed proposed by a user who tried none platform cluster?

File: internal/cluster/validations/validations.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielerez) [reply internal/cluster/validations/validations.go:38]: Yeah, commented here: https://github.com/openshift/assisted-service/pull/4778#issuecomment-1351629129

**Comment by CrystalChun** [comment]
> 
> Oh nice catch, so it's a limitation of the platforms then. So maybe we could just enable the subdomain requirement only for none platform for now? I.e. It could be safer to have a stricter validation to avoid this issue when adding more platforms later on. Do we know whether the subdomain requirement indeed proposed by a user who tried none platform cluster?

Good idea! I agree to enable the requirement just for None platform. 

The subdomain requirement seems like it was a bare metal platform ([slack thread](https://coreos.slack.com/archives/CUPJTHQ5P/p1671032421447629?thread_ts=1660243070.543409&cid=CUPJTHQ5P))

File: internal/cluster/validations/validations.go [hunk 2]
```diff
@@ -326,8 +327,8 @@ func getRegistriesWithAuth(ignoreList string, ignoreSeparator string, images ...
 	return &registries, nil
 }
 
-//ValidateVipDHCPAllocationWithIPv6 returns an error in case of VIP DHCP allocation
-//being used with IPv6 machine network
+// ValidateVipDHCPAllocationWithIPv6 returns an error in case of VIP DHCP allocation
+// being used with IPv6 machine network
```
[the remark this answers is not in the document]
Comment (danielerez) [reply internal/cluster/validations/validations.go:337]: Yeah preferably it should be done separately, but not crucial.

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -603,7 +603,7 @@ func (b *bareMetalInventory) RegisterClusterInternal(
 	}
 	setPullSecret(&cluster, ps)
 
-	if err = validations.ValidateClusterNameFormat(swag.StringValue(params.NewClusterParams.Name)); err != nil {
+	if err = validations.ValidateClusterNameFormat(swag.StringValue(params.NewClusterParams.Name), *params.NewClusterParams.Platform.Type); err != nil {
```
Comment (danielerez) [inline internal/bminventory/inventory.go:606]: iinm, it's nil-safe here, but would be cleaner to use 'getPlatformType' func: https://github.com/openshift/assisted-service/blob/be2c5764fd0f2d90750d650aff267cda579fa4ef/internal/bminventory/inventory.go#L1818

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -868,7 +868,7 @@ func (b *bareMetalInventory) V2ImportClusterInternal(ctx context.Context, kubeKe
 		KubeKeyNamespace: kubeKey.Namespace,
 	}
 
-	err := validations.ValidateClusterNameFormat(clusterName)
+	err := validations.ValidateClusterNameFormat(clusterName, *newCluster.Platform.Type)
```
Comment (danielerez) [inline internal/bminventory/inventory.go:871]: same, i.e. getPlatformType(newCluster.Platform.Type)

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (CrystalChun) [reply internal/bminventory/inventory.go]: Gotcha, modified to use this function, thanks!

File: internal/cluster/validations/validations.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (CrystalChun) [reply internal/cluster/validations/validations.go]: Ok good to know! Removed it from the PR. Thank you!

**Review by danielerez — COMMENTED** [review]
looks good:) just need to fix lint format issue here: https://github.com/openshift/assisted-service/pull/4778/files#diff-142e4e6d84ff9769edd6c1e72b908d885324d62da8048f25f24dcb8f86fbbb42R216-R217

**Comment by CrystalChun** [comment]
> looks good:) just need to fix lint format issue here: https://github.com/openshift/assisted-service/pull/4778/files#diff-142e4e6d84ff9769edd6c1e72b908d885324d62da8048f25f24dcb8f86fbbb42R216-R217

Thanks for catching! Fixed it!

**Pull request merged**